// node-firebird-driver/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlobStream = exports.BlobSeekWhence = exports.Blob = exports.TransactionIsolation = void 0;
/** TransactionIsolation enum */
var TransactionIsolation;
(function (TransactionIsolation) {
    TransactionIsolation["CONSISTENCY"] = "CONSISTENCY";
    TransactionIsolation["READ_COMMITTED"] = "READ_COMMITTED";
    TransactionIsolation["SNAPSHOT"] = "SNAPSHOT";
})(TransactionIsolation || (exports.TransactionIsolation = TransactionIsolation = {}));
/** Blob class. */
class Blob {
    constructor(attachment, id) {
        /** Gets the blob's id. */
        this.id = new Uint8Array(8);
        this.attachment = attachment;
        this.id.set(id);
    }
    /** True if the blob's attachment is valid. */
    get isValid() {
        return this.attachment.isValid;
    }
}
exports.Blob = Blob;
var BlobSeekWhence;
(function (BlobSeekWhence) {
    BlobSeekWhence[BlobSeekWhence["START"] = 0] = "START";
    BlobSeekWhence[BlobSeekWhence["CURRENT"] = 1] = "CURRENT";
    BlobSeekWhence[BlobSeekWhence["END"] = 2] = "END";
})(BlobSeekWhence || (exports.BlobSeekWhence = BlobSeekWhence = {}));
/** BlobStream class. */
class BlobStream {
    constructor(blob) {
        this.blob = blob;
    }
}
exports.BlobStream = BlobStream;
